File: examples/python/vault_client.py
"""
vault_client.py — Read and write secrets using HashiCorp Vault via hvac.

Prerequisites:
    pip install -r requirements.txt
    export VAULT_ADDR="https://vault.example.com:8200"
    export VAULT_TOKEN="<YOUR_VAULT_TOKEN>"  # or use vault login first

Usage:
    python vault_client.py
"""

import os
import sys
import hvac


def create_client() -> hvac.Client:
    """Create and authenticate a Vault client from environment variables."""
    vault_addr = os.environ.get("VAULT_ADDR", "http://127.0.0.1:8200")
    vault_token = os.environ.get("VAULT_TOKEN")

    client = hvac.Client(url=vault_addr, token=vault_token)

    if not client.is_authenticated():
        print("ERROR: Not authenticated to Vault.", file=sys.stderr)
        print("Set VAULT_ADDR and VAULT_TOKEN environment variables.", file=sys.stderr)
        sys.exit(1)

    return client


def write_secret(client: hvac.Client, path: str, data: dict) -> None:
    """Write a secret to Vault KV v2."""
    # path is relative to the KV mount, e.g. "myproject/db"
    client.secrets.kv.v2.create_or_update_secret(path=path, secret=data)
    print(f"✅ Secret written to: {path}")


def read_secret(client: hvac.Client, path: str) -> dict:
    """Read a secret from Vault KV v2. Returns the data dict."""
    response = client.secrets.kv.v2.read_secret_version(path=path, raise_on_deleted_version=True)
    return response["data"]["data"]


def read_field(client: hvac.Client, path: str, field: str) -> str:
    """Read a single field from a KV v2 secret."""
    data = read_secret(client, path)
    if field not in data:
        raise KeyError(f"Field '{field}' not found at path '{path}'")
    return data[field]


def list_secrets(client: hvac.Client, path: str) -> list[str]:
    """List secret keys at a path."""
    response = client.secrets.kv.v2.list_secrets(path=path)
    return response["data"]["keys"]


def delete_secret(client: hvac.Client, path: str) -> None:
    """Soft-delete the latest version of a secret."""
    client.secrets.kv.v2.delete_latest_version_of_secret(path=path)
    print(f"🗑️  Secret deleted (soft): {path}")


def demo() -> None:
    client = create_client()

    # Write
    write_secret(client, "myproject/demo", {
        "username": "app_user",
        "password": "<DEMO_PASSWORD>",
    })

    # Read
    data = read_secret(client, "myproject/demo")
    print(f"Read secret: username={data['username']}")

    # Read single field
    password = read_field(client, "myproject/demo", "password")
    print(f"Password field retrieved (length: {len(password)})")

    # List
    keys = list_secrets(client, "myproject/")
    print(f"Secrets at myproject/: {keys}")

    # Clean up
    delete_secret(client, "myproject/demo")


if __name__ == "__main__":
    demo()

File: examples/typescript/vault-client.ts
/**
 * vault-client.ts — Read and write secrets using HashiCorp Vault (node-vault).
 *
 * Prerequisites:
 *   npm install
 *   export VAULT_ADDR="https://vault.example.com:8200"
 *   export VAULT_TOKEN="<YOUR_VAULT_TOKEN>"
 *
 * Usage:
 *   npx ts-node vault-client.ts
 */

import nodeVault from "node-vault";

interface VaultClient {
  write(path: string, data: Record<string, unknown>): Promise<unknown>;
  read(path: string): Promise<{ data: { data: Record<string, string> } }>;
  list(path: string): Promise<{ data: { keys: string[] } }>;
  delete(path: string): Promise<unknown>;
}

function createClient(): VaultClient {
  const addr = process.env.VAULT_ADDR ?? "http://127.0.0.1:8200";
  const token = process.env.VAULT_TOKEN;

  if (!token) {
    console.error("ERROR: VAULT_TOKEN is not set.");
    process.exit(1);
  }

  return nodeVault({ apiVersion: "v1", endpoint: addr, token }) as VaultClient;
}

/**
 * Write a secret to Vault KV v2.
 * Path is relative to the KV mount, e.g. "myproject/db".
 */
async function writeSecret(
  client: VaultClient,
  mountPath: string,
  secretPath: string,
  data: Record<string, string>
): Promise<void> {
  await client.write(`${mountPath}/data/${secretPath}`, { data });
  console.log(`✅ Secret written to: ${mountPath}/data/${secretPath}`);
}

/**
 * Read a KV v2 secret and return its data object.
 */
async function readSecret(
  client: VaultClient,
  mountPath: string,
  secretPath: string
): Promise<Record<string, string>> {
  const response = await client.read(`${mountPath}/data/${secretPath}`);
  return response.data.data;
}

/**
 * Read a single field from a KV v2 secret.
 */
async function readField(
  client: VaultClient,
  mountPath: string,
  secretPath: string,
  field: string
): Promise<string> {
  const data = await readSecret(client, mountPath, secretPath);
  if (!(field in data)) {
    throw new Error(`Field "${field}" not found at ${mountPath}/${secretPath}`);
  }
  return data[field];
}

/**
 * List secret keys at a KV v2 path.
 */
async function listSecrets(
  client: VaultClient,
  mountPath: string,
  prefix: string
): Promise<string[]> {
  const response = await client.list(`${mountPath}/metadata/${prefix}`);
  return response.data.keys;
}

/**
 * Soft-delete the latest version of a KV v2 secret.
 */
async function deleteSecret(
  client: VaultClient,
  mountPath: string,
  secretPath: string
): Promise<void> {
  await client.delete(`${mountPath}/data/${secretPath}`);
  console.log(`🗑️  Secret deleted: ${mountPath}/${secretPath}`);
}

async function main(): Promise<void> {
  const client = createClient();
  const mount = "secret";
  const path = "myproject/demo";

  await writeSecret(client, mount, path, {
    username: "app_user",
    password: "<DEMO_PASSWORD>",
  });

  const data = await readSecret(client, mount, path);
  console.log(`Read secret: username=${data.username}`);

  const password = await readField(client, mount, path, "password");
  console.log(`Password field retrieved (length: ${password.length})`);

  const keys = await listSecrets(client, mount, "myproject/");
  console.log(`Secrets at myproject/: ${keys.join(", ")}`);

  await deleteSecret(client, mount, path);
}

main().catch((err) => {
  console.error("Fatal error:", err.message);
  process.exit(1);
});
